narwhal/config/src/committee.rs

```diff
@@ -0,0 +1,380 @@
+        self.authorities.iter()
+    }
+
+    pub fn authority_by_network_key(
+        &self,
+        network_key: &NetworkPublicKey,
+    ) -> Option<(&PublicKey, &Authority)> {
+        self.authorities
+            .iter()
+            .find(|(_, authority)| authority.network_key == *network_key)
+    }
+
+    /// Returns the number of authorities.
+    pub fn size(&self) -> usize {
+        self.authorities.len()
+    }
+
+    /// Return the stake of a specific authority.
+    pub fn stake(&self, name: &PublicKey) -> Stake {
+        self.authorities
+            .get(&name.clone())
+            .map_or_else(|| 0, |x| x.stake)
+    }
+
+    pub fn stake_by_id(&self, id: &AuthorityIdentifier) -> Stake {
+        self.authorities_by_id
+            .get()
+            .unwrap()
+            .get(id)
+            .map_or_else(|| 0, |authority| authority.stake)
+    }
+
+    /// Returns the stake required to reach a quorum (2f+1).
+    pub fn quorum_threshold(&self) -> Stake {
+        // If N = 3f + 1 + k (0 <= k < 3)
+        // then (2 N + 3) / 3 = 2f + 1 + (2k + 2)/3 = 2f + 1 + k = N - f
+        let total_votes: Stake = self.authorities.values().map(|x| x.stake).sum();
+        2 * total_votes / 3 + 1
+    }
+
+    /// Returns the stake required to reach availability (f+1).
+    pub fn validity_threshold(&self) -> Stake {
+        // If N = 3f + 1 + k (0 <= k < 3)
+        // then (N + 2) / 3 = f + 1 + k/3 = f + 1
+        let total_votes: Stake = self.authorities.values().map(|x| x.stake).sum();
+        (total_votes + 2) / 3
+    }
+
+    /// Returns a leader node as a weighted choice seeded by the provided integer
+    pub fn leader(&self, seed: u64) -> PublicKey {
+        let mut seed_bytes = [0u8; 32];
+        seed_bytes[32 - 8..].copy_from_slice(&seed.to_le_bytes());
+        let mut rng = StdRng::from_seed(seed_bytes);
+        let choices = self
+            .authorities
+            .iter()
+            .map(|(name, authority)| (name, authority.stake as f32))
+            .collect::<Vec<_>>();
+        choices
+            .choose_weighted(&mut rng, |item| item.1)
+            .expect("Weighted choice error: stake values incorrect!")
+            .0
+            .clone()
+    }
+
+    /// Returns the primary address of the target primary.
+    pub fn primary(&self, to: &PublicKey) -> Result<Multiaddr, ConfigError> {
+        self.authorities
+            .get(&to.clone())
+            .map(|x| x.primary_address.clone())
+            .ok_or_else(|| ConfigError::NotInCommittee((*to).encode_base64()))
+    }
+
+    pub fn network_key(&self, pk: &PublicKey) -> Result<NetworkPublicKey, ConfigError> {
+        self.authorities
+            .get(&pk.clone())
+            .map(|x| x.network_key.clone())
+            .ok_or_else(|| ConfigError::NotInCommittee((*pk).encode_base64()))
+    }
+
+    /// Return all the network addresses in the committee.
+    pub fn others_primaries(
```

Comment:
ooc, do we invoke these again and again? Maybe I'm missing something here, but I thought the whole idea is in each epoch we should have a Hashmap of valid committee members (where key = index - the identifier we use in this PR). Then we always ask that hashmap to avoid clones, extra serializations/deserializations etc, is it still the case?

Reply:
> Maybe I'm missing something here, but I thought the whole idea is in each epoch we should have a Hashmap of valid committee members (where key = index - the identifier we use in this PR)

@kchalkias that's what we do [here](https://github.com/MystenLabs/sui/pull/9576/files#diff-bf39a5ebab429076e198116328c16f4d2d6176154e4d649250ff9cc00e93c8c3R97) and we leverage whenever we do look ups by the specified id

> When the epoch advances we update that map with the new committee members (and I guess that map could be reused between Sui and NW

Yep, as we always shutdown/start a fresh Narwhal we re-create the Committee with the new authorities

> Actually if it's an index, then we just need a vector, not even a Hashmap as its index-in-the-vector will hold a reference to the corresponding committee member.

We could absolutely do that - I've consciously used a `BTreeMap` to  keep the implementation more generic and independent of the nature of the `AuthorityIdentifier` - I don't consider a huge opportunity cost of a look up in a BTreeMap of `100` elements Vs a `Vec` - but we can totally convert to the latter if we want to super optimise things.

Reply:
We can cache this inside Committee, but in making Committee larger, we would need to avoid excessive cloning of it too.